File: src/python/_0238_Product_of_Array_Except_Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [1] * len(nums)

        for i in range(1, len(nums)):
            prefix[i] = prefix[i - 1] * nums[i - 1]

        suffix = 1
        for i in range(len(nums) - 1, -1, -1):
            prefix[i] *= suffix
            suffix *= nums[i]

        return prefix

File: src/_0238_Product_of_Array_Except_Self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] prefix = new int[nums.length];
        prefix[0] = 1;

        for (int i = 1; i < nums.length; i++)
            prefix[i] = prefix[i - 1] * nums[i - 1];

        int suffix = 1;
        for (int i = nums.length - 1; i >= 0; suffix *= nums[i--])
            prefix[i] *= suffix;

        return prefix;
    }
}
